src/dag.rs

```diff
@@ -31,42 +39,329 @@ impl<Types: RepoTypes> IpldDag<Types> {
-        Ipld::List(mut vec) => match segment.parse::<usize>() {
-            Ok(index) if index < vec.len() => Ok(vec.swap_remove(index)),
-            Ok(_) => Err(anyhow::anyhow!(
-                "index out of range 0..{}: {:?}",
-                vec.len(),
-                segment
-            )),
-            Err(_) => Err(anyhow::anyhow!("invalid list index: {:?}", segment)),
-        },
-        link @ Ipld::Link(_) if segment == "." => Ok(link),
-        other => Err(anyhow::anyhow!(
-            "cannot resolve {:?} through: {:?}",
-            segment,
-            other
-        )),
+impl TryFrom<ResolvedNode> for Ipld {
+    type Error = Error;
+    fn try_from(r: ResolvedNode) -> Result<Ipld, Error> {
+        use ResolvedNode::*;
+
+        match r {
+            Block(block) => Ok(decode_ipld(block.cid(), block.data())?),
+            DagPbData(_, node_data) => Ok(Ipld::Bytes(node_data.node_data().to_vec())),
+            Projection(_, ipld) => Ok(ipld),
+            Link(_, cid) => Ok(Ipld::Link(cid)),
+        }
+    }
+}
+
+/// Success variants for resolve_local operation on an Ipld document.
+#[derive(Debug)]
+enum LocallyResolved<'a> {
+    /// Resolution completed
+    Complete(ResolvedNode),
+
+    /// Resolving was attempted on a block which is a HAMT sharded bucket, and needs to be
+    /// continued by loading other buckets.
+    Incomplete(Cid, ShardedLookup<'a>),
+}
+
+impl From<ResolvedNode> for LocallyResolved<'static> {
+    fn from(r: ResolvedNode) -> LocallyResolved<'static> {
+        LocallyResolved::Complete(r)
+    }
+}
+
+/// Resolves the given path segments locally or inside the given document; in addition to
+/// `resolve_local_ipld` this fn also handles normal dag-pb and unixfs HAMTs.
+fn resolve_local<'a>(
+    block: Block,
+    segments: &mut Peekable<impl Iterator<Item = &'a str>>,
+) -> Result<(LocallyResolved<'a>, usize), Error> {
+    println!("resolving on {}", block.cid());
+    if segments.peek().is_none() {
+        return Ok((LocallyResolved::Complete(ResolvedNode::Block(block)), 0));
+    }
+
+    let Block { cid, data } = block;
+
+    if cid.codec() == cid::Codec::DagProtobuf {
+        // converting dag-pb to ipld is quite tedious, not to mention HAMT and other stuff
+        // though, we cannot handle HAMT locally
+
+        let segment = segments.next().unwrap();
+
+        // TODO: lift the cache as parameter?
+        match resolve(&data, segment, &mut None) {
+            Ok(MaybeResolved::NeedToLoadMore(lookup)) => {
+                Ok((LocallyResolved::Incomplete(cid, lookup), 0))
+            }
+            Ok(MaybeResolved::Found(dest)) => {
+                // it's clear that we need to have the segment dropped here.
+                Ok((LocallyResolved::Complete(ResolvedNode::Link(cid, dest)), 1))
+            }
+            Ok(MaybeResolved::NotFound) => {
+                // TROUBLE: cannot honor the "pop only matched ones" anymore
```

Comment:
this trouble was solved by returning the usize and letting go what the iterator means. I think this also means that the iterator *should* be recreated for every block. Well, this case here terminates always.